quixstreams/sinks/community/mongodb.py

```diff
@@ -0,0 +1,210 @@
+            value["__partition"] = partition
+            value["__offset"] = record.offset
+        return value
+
+    def write(self, batch: SinkBatch) -> None:
+        """
+        Note: Transactions could be an option here, but then each record requires a
+        network call, and the transaction has size limits...so `bulk_write` is used
+        instead, with the downside that duplicate writes may occur if errors arise.
+        """
+        start = time.monotonic()
+        min_timestamp = sys.maxsize
+        max_timestamp = -1
+        doc_match = self._document_matcher
+        update_method = self._update_method
+        upsert = self._upsert
+        value_selector = self._value_selector
+        is_update = self._update_method != ReplaceOne
+
+        if self._add_message_metadata or self._add_topic_metadata:
+            added_metadata = partial(self._add_metadata, batch.topic, batch.partition)
+        else:
+            added_metadata = None
+
+        records = []
+        for record in batch:
+            ts = record.timestamp
+            value = added_metadata(record) if added_metadata else record.value
+            value = value_selector(value) if value_selector else value
+            records.append(
+                update_method(
+                    doc_match(record) if doc_match else None,
+                    {"$set": value} if is_update else value,
+                    upsert=upsert,
+                )
+            )
+            min_timestamp = min(ts, min_timestamp)
+            max_timestamp = max(ts, max_timestamp)
+
+        attempts_remaining = 3
+        while attempts_remaining:
+            try:
+                # We don't need to worry about manually batching our writes, the MongoDB
+                # client does it for us automatically (chunks every 10k records).
+                # It will block until all writes are successful.
+                self._collection.bulk_write(records)
+                logger.info(
+                    "Sent update requests to MongoDB; "
+                    f"messages_processed={batch.size} "
+                    f"host_address={self._client.address[0]} "
+                    f"database={self._db_name} "
+                    f"collection={self._collection_name} "
+                    f"min_timestamp={min_timestamp} "
+                    f"max_timestamp={max_timestamp} "
+                    f"time_elapsed={round(time.monotonic() - start, 2)}s"
+                )
+                return
+            except (NetworkTimeout, ExecutionTimeout, WriteConcernError) as e:
+                logger.error(f"MongoDB: Encountered a network or server error: {e}")
+            attempts_remaining -= 1
+            if attempts_remaining:
+                backoff = 5
+                logger.error(
+                    f"Sleeping for {backoff} seconds and re-attempting. "
+                    f"Attempts remaining: {attempts_remaining}"
+                )
+                time.sleep(backoff)
```

Comment:
1. I'd move it to except block, otherwise it is slightly confusing.
2. You can decrement `attempts_remaining` at the end to avoid `if`.
3. There are two backoff constants, maybe you can set those in the module scope as `ATTEMPT_BACKOFF` and `RETRY_BACKOFF`.
```suggestion
                logger.error(
                    f"Sleeping for {ATTEMPT_BACKOFF} seconds and re-attempting. "
                    f"Attempts remaining: {attempts_remaining}"
                )
                time.sleep(ATTEMPT_BACKOFF)
                attempts_remaining -= 1
```

Reply:
Yeah, I was torn on a little on the layout here. Indent suggestion seems reasonable and I originally had it that way before changing it to this :joy: 

The `if` is still useful here since it means not sleeping an extra `ATTEMPT_BACKOFF` seconds when it's not supposed to attempt anymore...in the grand scheme of things it probably wont matter, but that was my reasoning for doing the check.